src/app/clusters/bindings/BindingManager.cpp

```diff
@@ -0,0 +1,192 @@
+    });
+}
+
+void BindingManager::SyncPendingClustersToPeer(OperationalDeviceProxy * device, PendingClusterEntry * pendingClusters)
+{
+    for (uint8_t i = 0; i < pendingClusters->mNumPendingClusters; i++)
+    {
+        ClusterId cluster   = pendingClusters->mPendingClusters[i].cluster;
+        EndpointId endpoint = pendingClusters->mPendingClusters[i].endpoint;
+        for (uint8_t j = 0; j < EMBER_BINDING_TABLE_SIZE; j++)
+        {
+            EmberBindingTableEntry entry;
+            if (emberGetBinding(j, &entry) == EMBER_SUCCESS && entry.type != EMBER_UNUSED_BINDING && entry.clusterId == cluster &&
+                entry.local == endpoint && mBoundDeviceChangedHandler)
+            {
+                mBoundDeviceChangedHandler(entry.local, entry.remote, cluster, device);
+            }
+        }
+    }
+    mPendingClusterMap.RemoveEntry(pendingClusters);
+}
+
+void BindingManager::HandleDeviceConnectionFailure(void * context, NodeId nodeId, CHIP_ERROR error)
+{
+    BindingManager * manager = static_cast<BindingManager *>(context);
+    manager->HandleDeviceConnectionFailure(nodeId, error);
+}
+
+void BindingManager::HandleDeviceConnectionFailure(NodeId nodeId, CHIP_ERROR error)
+{
+    // Simply release the entry, the connection will be re-established on need.
+    ChipLogError(AppServer, "Failed to establish connection to node 0x" ChipLogFormatX64, ChipLogValueX64(nodeId));
+    mAppServer->GetCASESessionManager()->ReleaseSession(nodeId);
+}
+
+CHIP_ERROR BindingManager::DisconnectDevice(FabricIndex fabric, NodeId node)
+{
+    FabricInfo * fabricInfo     = mAppServer->GetFabricTable().FindFabricWithIndex(fabric);
+    PendingClusterEntry * entry = mPendingClusterMap.FindEntry(fabricInfo->GetPeerIdForNode(node));
+
+    VerifyOrReturnError(entry != nullptr, CHIP_ERROR_NOT_FOUND);
+
+    mAppServer->GetCASESessionManager()->ReleaseSession(node);
+    mPendingClusterMap.RemoveEntry(entry);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BindingManager::NotifyBoundClusterChanged(EndpointId endpoint, ClusterId cluster)
+{
+    for (uint8_t i = 0; i < EMBER_BINDING_TABLE_SIZE; i++)
+    {
+        EmberBindingTableEntry entry;
+        CHIP_ERROR error = CHIP_NO_ERROR;
+
+        if (emberGetBinding(i, &entry) == EMBER_SUCCESS && entry.type != EMBER_UNUSED_BINDING && entry.local == endpoint &&
+            entry.clusterId == cluster)
+        {
+            OperationalDeviceProxy * peerDevice = mAppServer->GetCASESessionManager()->FindExistingSession(entry.nodeId);
+            if (peerDevice != nullptr && mBoundDeviceChangedHandler)
+            {
+                // We already have an active connection
+                mBoundDeviceChangedHandler(entry.local, entry.remote, cluster, peerDevice);
+            }
+            else
+            {
+                // Enqueue pending cluster and establish connection
+                error = EnqueueClusterAndConnect(entry.fabricIndex, entry.nodeId, entry.local, entry.clusterId);
```

Comment:
are we supposed to return the error on failure?

it looks like this method always returns CHIP_NO_ERROR regardless if underlying methods fail or succeed.